File: source/modulo_core/src/JointPositionsListener.cpp
#include "modulo_core/JointPositionsListener.hpp"

#include "modulo_core/exceptions.hpp"

namespace modulo_core {

static state_representation::JointPositions
translate_joint_positions(const modulo_interfaces::msg::JointPositions& joint_positions) {
  auto state = state_representation::JointPositions(joint_positions.header.frame_id, joint_positions.joint_names);
  state.set_positions(joint_positions.positions);
  return state;
}

JointPositionsListener::~JointPositionsListener() {
  if (spin_thread_) {
    executor_->cancel();
    dedicated_listener_thread_->join();
  }
}

void JointPositionsListener::subscription_callback(
    const std::shared_ptr<modulo_interfaces::msg::JointPositionsCollection> message) {
  for (const auto& msg : message->joint_positions) {
    buffer_.push_back(msg);
  }
}

template<>
modulo_interfaces::msg::JointPositions JointPositionsListener::lookup(const std::string& name) const {
  auto result =
      std::find_if(buffer_.begin(), buffer_.end(), [name](modulo_interfaces::msg::JointPositions joint_positions) {
        return joint_positions.header.frame_id == name;
      });
  if (result == buffer_.end()) {
    throw exceptions::LookupJointPositionsException("Joint positions " + name + " is not available");
  }
  return *result;
}

template<>
state_representation::JointPositions JointPositionsListener::lookup(const std::string& name) const {
  return translate_joint_positions(lookup<modulo_interfaces::msg::JointPositions>(name));
}

template<>
std::vector<modulo_interfaces::msg::JointPositions> JointPositionsListener::get_buffer() const {
  return buffer_;
}

template<>
std::vector<state_representation::JointPositions> JointPositionsListener::get_buffer() const {
  std::vector<state_representation::JointPositions> states;
  for (const auto& joint_positions : buffer_) {
    states.push_back(translate_joint_positions(joint_positions));
  }
  return states;
}
}// namespace modulo_core

File: source/modulo_core/modulo_core/joint_positions_listener.py
from threading import Thread
from typing import List, Optional, Union

from modulo_core.exceptions import LookupJointPositionsException
from modulo_interfaces.msg import JointPositions, JointPositionsCollection
from rclpy.callback_groups import ReentrantCallbackGroup
from rclpy.executors import SingleThreadedExecutor
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy, HistoryPolicy, QoSProfile


class JointPositionsListener:
    """
    The JointPositionsBroadcaster is a TF2 style class that listens to the fixed /joint_positions topic and allows to
    lookup messages from a buffer.
    """

    def __init__(
            self, node: Node, spin_thread: bool = False, qos: Optional[Union[QoSProfile, int]] = None):
        """
        Constructor of the JointPositionsListener.

        :param node: The ROS2 node
        :param spin_thread: Whether to create a dedicated thread to spin this node
        :param qos: A QoSProfile or a history depth to apply to the subscriber
        """
        if qos is None:
            qos = QoSProfile(depth=100, durability=DurabilityPolicy.TRANSIENT_LOCAL, history=HistoryPolicy.KEEP_LAST)
        self._node = node
        # Default callback group is mutually exclusive, which would prevent waiting for joint positions
        # from another callback in the same group.
        self._group = ReentrantCallbackGroup()
        self._buffer = []

        self._subscription = node.create_subscription(
            JointPositionsCollection, "/joint_positions", self._callback, qos, callback_group=self._group)

        if spin_thread:
            self._executor = SingleThreadedExecutor()

            def run_func():
                self._executor.add_node(self._node)
                self._executor.spin()
                self._executor.remove_node(self._node)

            self._dedicated_listener_thread = Thread(target=run_func)
            self._dedicated_listener_thread.start()

    def __del__(self):
        if hasattr(self, 'dedicated_listener_thread') and hasattr(self, 'executor'):
            self._executor.shutdown()
            self._dedicated_listener_thread.join()
        self._node.destroy_subscription(self._subscription)

    def _callback(self, message: JointPositionsCollection):
        for msg in message.joint_positions:
            self._buffer.append(msg)

    def lookup(self, name: str) -> JointPositions:
        """
        Look up JointPositions object by it's name.

        :param name: The name of the JointPositions object to lookup
        """
        for joint_positions in self._buffer:
            if joint_positions.header.frame_id == name:
                return joint_positions
        raise LookupJointPositionsException(f"Joint positions {name} is not available")

    @property
    def buffer(self) -> List[JointPositions]:
        """
        Get a vector of all buffered JointPositions objects.
        """
        return self._buffer
